---
title: "Preliminary Analyses: Immune-checkpoint blockade (ICB) data"
#author: "BHK lab"
date: '2022-09-07'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(ggplot2)
library(survival)
library(survminer)
library(readr)
library(magrittr)
library(summarytools)

```

## Load data

To download and extract TSV version of ICB data from Zenodo (https://zenodo.org/record/7007756/files/), follow the script named download_tsv_data R file at "~/Code/Download_Data".  

As an example, we consider the ICB Braun data with PMID 32472114 and read RNA-seq expression, clinicopathological characteristics, and gene meta data. The Braun study includes 319 patients with clinical data and 40,994 genes. While only 181 patients have both clinical data and RNA-seq expression data. 

Note that the expression data is $log2(\text{TPM} + 0.001)$, where TPM represents transcripts per million. 

```{r load ICB data}

# clinicopathological characteristics  
#dat_sample = read.table('C:/IOProject/Roche/ICB_Braun_coldata.tsv', sep="\t") %>% as.data.frame()
dat_sample = read.table('~/Data/ICB_Braun/ICB_Braun_metadata.tsv', sep="\t") %>% as.data.frame()
dim(dat_sample)

# gene meta data
dat_gene = read.table('~/Data/ICB_Braun/ICB_Braun_expr_genes.tsv', sep="\t") %>% as.data.frame()
dim(dat_gene)

# RNA-seq expression data (i.e., log2(TPM + 0.001)) 
dat_exp = read.table('~/Data/ICB_Braun/ICB_Braun_expr.tsv', sep="\t") %>% as.data.frame()
dim(dat_exp)

# Get the subset of data with both clinical and expression data

dat_sample = dat_sample[rownames(dat_sample) %in% colnames(dat_exp), ]

dim(dat_sample)

# Order the samples in both clinical and expression data

dat_sample <- dat_sample[order(rownames(dat_sample)), ]
dat_exp <- dat_exp[, order(colnames(dat_exp))]

```

## Clinical data summary

The detailed clinicopathological characteristics of the Braun cohort including cancer type, age, sex, respose, overall survival (OS, "t.os"), and progression-free survival (PFS, "survival_time").

```{r clinical data, results='asis'}
dat_sample_sub <- dat_sample[, c("age","sex", "response","t.os", "survival_time")]

#view(dfSummary(dat_sample_sub))

print(dfSummary(dat_sample_sub, 
                style = "grid", 
                plain.ascii = FALSE, 
                graph.col = FALSE), 
      method = 'render')

```


We assess under the null hypothesis that the sex and response is independent using the Chi-squared test. The p-value = 0.96 shows we don't have enough evidence to reject the null hypothesis.

```{r sex response, results='asis'}

dat_sample_sub <- dat_sample_sub[!is.na(dat_sample_sub$response), ]
dat_sample_sub %$%  
  ctable(x = sex, y = response,
         chisq = TRUE,
         OR    = TRUE,
         RR    = TRUE,
         headings = FALSE) %>%
  print(method = "render")

```

## Logistic regression analysis

To assess the association between a gene and response (R vs. NR) variable, we fit the logistic regression model. The structural form of the logistic regression model is given as

$$\textit{logit}(E(Y|X)) = \textit{logit}(p) = \textit{ln} (\frac{p}{1-p}) = \beta_0 + \beta_1 X + \epsilon$$
where

* $Y$  denotes the outcome or dependent variable (e.g., response variable); this is a binary variable $X$  denotes the predictor of interest or the independent variable (e.g., expression data)
* $p$ denotes the probability of the event occurring, and $\frac{p}{1-p}$ shows the odds ratio (OR)
* $\beta_0$ denotes the Y-intercept when X is zero; this is not informative for logistic regression models
* $\beta_1$ denotes the slope or the expected change in log odds per unit change in X
* $\epsilon$ denotes the errors

The logistic regression model has several assumptions such as observations are independent of each other, no multi-collinearity across independent variables, and linearity of log odds. 

For a given gene ENSG00000065491.8 (TBC1D22B), the boxplot shows 22\% of responses are missing. After removing the missing values, the fitted logistic model represents that there is significant association between gene and response ($log\text{OR} = 0.98$ and $p = 0.003$).


```{r logit single gene}

# for a given gene: ENSG00000065491.8

df <- data.frame(x = as.numeric(dat_exp[rownames(dat_exp) == "ENSG00000065491.8",]),
                 t.os = dat_sample$t.os,
                 os = dat_sample$os,
                 sex = dat_sample$sex,
                 y = dat_sample$response)


# boxplot to show the distribution of expression across response groups (R vs. NR)

p <- ggplot(data=df, aes(x= y, y= x, 
                          fill=y)) + 
  geom_boxplot() + 
  ylab("gene expression") +
  xlab("") +
  theme(axis.text.x=element_text(size=12,  face="bold"),
        axis.title=element_text(size=14,face="bold"),
        axis.text.y=element_text(size=12, face = "bold"),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        plot.background = element_blank(), 
        axis.line = element_line(colour = "black"),
        legend.position="bottom",
        legend.text = element_text(size = 8, face="bold"),
        legend.title = element_blank())

p

# Before fitting the logistic regression model, the missing values are removed

df <- df[!is.na(df$y),]
df$y <- factor(df$y)
df$y <- ifelse(df$y == "R", 0, 1)

fit <- glm( y ~ x, 
              data = df, 
              family = "binomial")

fit_summary <- summary(fit)
fit_summary

data.frame(gene_id = "ENSG00000065491.8",
           "logOR" = fit_summary$coefficients["x", "Estimate"],
           "SE" = fit_summary$coefficients["x", "Std. Error"],
           "pvals" = fit_summary$coefficients["x", "Pr(>|z|)"])

```


## Survival analysis

### Part I

We consider the Kaplan-Meier (KM) and log-rank test to assess if there is any statistically difference in survival rate (e.g., OS) if we divide our data based on sex (F vs. M).

The log-rank test $p = 0.11$ shows there is not any significant difference in survival rate across two group of sex (F vs. M).  

```{r KM sex}

df <- data.frame(x = as.numeric(dat_exp[rownames(dat_exp) == "ENSG00000065491.8",]),
                 t.os = dat_sample$t.os,
                 os = dat_sample$os,
                 sex = dat_sample$sex,
                 y = dat_sample$response)

surv_object <- Surv(time = df$t.os, event = df$os)
fit <- survfit(surv_object ~ sex, data = df)

ggsurvplot(fit, data = df, pval = TRUE, risk.table = TRUE)

```

We consider the KM and log-rank test to assess if there is any statistically difference in survival rate (e.g., OS) if we divide our data based on response (R vs. NR).

The log-rank test $p < 0.0001$ shows there is considerable difference in survival rate across two group of response (R vs. NR).  		


```{r KM response}


surv_object <- Surv(time = df$t.os, event = df$os)
fit <- survfit(surv_object ~ y, data = df)

ggsurvplot(fit, data = df, pval = TRUE, risk.table = TRUE)

```


### Part II

To assess the association between genes and time-to-event variables (e.g., OS), we fit the Cox proportional hazards model. The semi-parametric Cox PH model is given as 

$$h(t) = h_0(t) \times \text{exp}(\beta X)$$
where 

* $t$ represents the survival time.
* $h(t)$ is the hazard function determined by a set of covariate (e.g., X)
* The coefficient $\beta$ measures the impact of covariate.
* The term $h_0$ is called the baseline hazard. It corresponds to the value of the hazard if the covariate is equal to zero. The $t$ in $h(t)$ reminds us that the hazard may vary over time.
* The quantities $\text{e}^\beta$ are called hazard ratios (HR).

The fitted Cox model with $p = 0.52$ shows no statistically significant association between OS and gene expression. 


```{r coxph}

fit <- coxph(Surv(t.os, os) ~ x, data = df)
fit_summary <- summary(fit)

fit_summary

data.frame(gene_id = "ENSG00000065491.8",
           "logHR" = fit_summary$coefficients["x", "coef"],
           "SE" = fit_summary$coefficients["x", "se(coef)"],
           "pvals" = fit_summary$coefficients["x", "Pr(>|z|)"])


```

