---
title: "ICB Data Curation Workflow"
author: "Minoru Nakano, Farnoosh Babazadeh, Sisira Nair"
date: "`r format(Sys.time(), '%a/%d/%b')`"
output: html_document
---

This RMarkdown script shows an example of ICB dataset curation workflow starting from curation to end analysis.\

## Introduction
### Data Curation Overview
We take data from published studies and curated them into standardized data objects to be used for integrated data analyses.
```{r echo=FALSE, out.width="50%"}
knitr::include_graphics("images/data_curation_overview.png")
```

**Standardized data curation resolves inconsistencies across different datasets:**\

* Various data formats: txt, csv, tsv, Excel, GEO...
* Various identifiers: "patientid" = "patient-id" = "id" = "patient" = "unique_patient_id"
* Unfriendly format: Newline characters, spaces, other unconventional characters
* Multiple data parts: patient/sample metadata, multiple assay data (RNA-seq, SNV...)

**The standardized data are integrated into MultiAssayExperiment object:**\
\
```{r echo=FALSE}
knitr::include_graphics("images/P2 - 30 min Workshop - Research camp.jpg")
```

MultiAssayExperiment:\

1. An R data object structure designed to hold multiple assay data for a set of patients/samples.
2. Assay data can be in other R data structures such as SummarizedExperiment.
3. Flexible: The data structure can handle assays with different number of observations.
4. Comes with functions to work with the data.

### Data Curation Process
```{r echo=FALSE, out.width="50%"}
knitr::include_graphics("images/data_curation_steps.png")
```

### ORCESTRA: Platform to Manage Data Curation Pipeline Executions
The ICB dataset curation pipelines were integrated and executed using the platform.\
```{r echo=FALSE, out.width="50%"}
knitr::include_graphics("images/data_curation_orcestra.png")
```

1. Each pipeline run is recorded and executed with a version controlled scripts.
2. Uses Snakemake to manage each step in a pipeline execution.
3. Establishes a clear association between the data and the version of code used to generate the data.
4. Provides a web interface to increase data accessibility. [www.orcestra.ca](https://www.orcestra.ca/)

## Data Curation Example: VanDenEnde
The scripts uses ICB_VanDenEnde data curation pipeline as an example. The rest of the datasets follow the same workflow.\
The pipeline is available in: [ICB_VanDenEnde-snakemake](https://github.com/BHKLAB-DataProcessing/ICB_VanDenEnde-snakemake)\
There are common functions used across all ICB dataset curation pipelines which are available in [ICB_Common](https://github.com/BHKLAB-DataProcessing/ICB_Common)\
\
The pipeline is organized in 4 steps:\

1. Downloads data from the publication source
2. Formats the downloaded data
3. Processes the formatted data into common data format across all ICB datasets
4. Creates MultiAssayExperiment object

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r message=FALSE}
library(data.table)
library(GEOquery)
library(Biobase)
library(SummarizedExperiment)
library(MultiAssayExperiment)
library(GenomicRanges)
library(biomaRt)
library(stringr)
library(readxl)
library(tximport)
```

```{r VanDenEnde study}
options(timeout=1000)

#' Before run, please add folder - Data and sub folders : download, annotation, processed, RDS and TSV
study <- 'VanDenEnde'
download_dir <- '~/Data/download'
annotation_dir <- '~/Data/annotation'
processed_dir <- '~/Data/processed'
rds_dir <- '~/Data/RDS'
tsv_dir <- '~/Data/TSV'
```

### 1. Download Source and Annotation Data
Publication: [https://pubmed.ncbi.nlm.nih.gov/33504550/](https://pubmed.ncbi.nlm.nih.gov/33504550/)\
Data source: [https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE165252](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE165252)

```{r VanDenEnde download data}

#' 1.1 Download source data
download.file(
  url='https://ftp.ncbi.nlm.nih.gov/geo/series/GSE165nnn/GSE165252/matrix/GSE165252_series_matrix.txt.gz',
  destfile = file.path(download_dir, 'GSE165252_series_matrix.txt.gz')
)
download.file(
  url='https://ftp.ncbi.nlm.nih.gov/geo/series/GSE165nnn/GSE165252/suppl/GSE165252_norm.cnt_PERFECT.txt.gz',
  destfile = file.path(download_dir, 'GSE165252_norm.cnt_PERFECT.txt.gz')
)

#' Download annotation files
#' Gene annotation data is kept in https://github.com/BHKLAB-DataProcessing/Annotations
#' Tissue and drug annotations for the ICB data are in https://github.com/BHKLAB-DataProcessing/ICB_Common/tree/main/data

#' 1.2 Download gene annotation data
download.file(
  url='https://github.com/BHKLAB-DataProcessing/Annotations/blob/master/Gencode.v19.annotation.RData?raw=true',
  destfile = file.path(annotation_dir, 'Gencode.v19.annotation.RData')
)

#' 1.3 Download tissue and drug annotation data
download.file(
  url='https://github.com/BHKLAB-DataProcessing/ICB_Common/raw/main/data/curation_drug.csv',
  destfile = file.path(annotation_dir, 'curation_drug.csv')
)
download.file(
  url='https://github.com/BHKLAB-DataProcessing/ICB_Common/raw/main/data/curation_tissue.csv',
  destfile = file.path(annotation_dir, 'curation_tissue.csv')
)
```


### 2. Format downloaded source data. 
This step extracts clinical and expression data from the source data format and parses them into more processing-friendly format.
```{r VanDenEnde format downloaded data}
commandArgs <- function(trailingOnly = TRUE){ return(c(download_dir)) }
source('https://raw.githubusercontent.com/BHKLAB-DataProcessing/ICB_VanDenEnde-snakemake/main/scripts/format_downloaded_data.R')
```
In the end, it produces two files extracted from the download source data: CLIN.txt and EXPR.txt.gz.

### 3. Process the formatted data to common data format across all ICB datasets
```{r VanDenEnde process the formatted data}
commandArgs <- function(trailingOnly = TRUE){ return(c(download_dir, processed_dir, annotation_dir)) }
source('https://raw.githubusercontent.com/BHKLAB-DataProcessing/ICB_VanDenEnde-snakemake/main/scripts/Format_Data.R')
```
In the end, it produces three files:\

1. CLIN.csv,
2. EXPR.csv and
3. cased_sequenced.csv, a file that indicates which patient has which molecular profiles. In this case, all patients will have EXPR data only.\

### 4. Create MultiAssayExperiment object
```{r VanDenEnde create MultiAssayExp object}
load(file.path(annotation_dir, "Gencode.v19.annotation.RData"))
source("https://raw.githubusercontent.com/BHKLAB-DataProcessing/ICB_Common/main/code/get_MultiAssayExp.R")
multi_assay_exp <- get_MultiAssayExp(study = study, input_dir = processed_dir)
```

**Data object is uploaded to Zenodo and made avaialble on ORCESTRA platform**: [ICB Datasets on ORCESTRA](https://www.orcestra.ca/clinical_icb)

### Using a MultiAssayExperiment Object
**Get a list of experiments included in this object**
```{r VanDenEnde using MultiAssayExp object 1}
assays(multi_assay_exp)
```
\
**Get patient metadata**
```{r VanDenEnde using MultiAssayExp object 2}
colData(multi_assay_exp)[, c('patientid', 'cancer_type', 'response', 'treatment')]
```
\
**View EXPR data assay**
```{r VanDenEnde using MultiAssayExp object 3}
expr <- experiments(multi_assay_exp)[['expr']]
assays(expr)[['expr']]
```

**View Gene Metadata for the Genes in the EXPR assay**
```{r VanDenEnde using MultiAssayExp object 4}
rowData(expr)
```

**Subsetting Data**
MultiAssayExperiment object can be subset by gene ids, patient data values and assay names.\
\
For exmaple: 
```{r VanDenEnde using MultiAssayExp object 5}
subset <- multi_assay_exp[c('ENSG00000003056.3', 'ENSG00000005073.5'), multi_assay_exp$response == 'R' & !is.na(multi_assay_exp$response),'expr']
colData(subset)[, c('patientid', 'cancer_type', 'response', 'treatment')]
```

View the subset EXPR data:
```{r VanDenEnde using MultiAssayExp object 6}
subset_expr <- experiments(subset)[['expr']]
assays(subset_expr)[['expr']]
```

View the subset gene metadata:
```{r VanDenEnde using MultiAssayExp object 7}
rowData(subset_expr)
```

### 5. Extract TSV files from the Data Object
\
RDS objects are downloaded and converted into TSV files using rds_to_tsv.R.\
They are available on [Zenodo](https://doi.org/10.5281/zenodo.7058399)\
\
Each zip file contains

1. patient metadata ICB_[Study]_metadata.tsv
2. assay data ICB_[Study]_[Assay Name].tsv
3. assay gene metadata ICB_[Study]_[Assay Name]_genes.tsv

The Assay Name can be "expr", "expr_gene_tpm", "expr_gene_counts", "expr_isoform_tpm", "expr_isoform_counts".

**"expr" and "expr_gene_tpm" are gene expression data in TPM, which is present across all datasets.**
```{r}
assay_name <- 'expr'
obj_name = paste0('ICB_', study)
dir.create(file.path(tsv_dir, obj_name))

#' output the clinical metadata
coldata <- data.frame(colData(multi_assay_exp))
write.table(
  coldata,
  file=file.path(tsv_dir, obj_name, paste0(obj_name, '_metadata', '.tsv')),
  sep='\t',
  col.names=TRUE,
  row.names=TRUE
)

#' output expr data and gene metadata
experiment <- experiments(multi_assay_exp)[[assay_name]]
assay <- data.frame(assay(experiment))
assay_genes <- data.frame(rowRanges(experiment))
write.table(
  assay,
  file=file.path(tsv_dir, obj_name, paste0(obj_name, '_', assay_name, '.tsv')),
  sep='\t',
  col.names=TRUE,
  row.names=TRUE
)
write.table(
  assay_genes,
  file=file.path(tsv_dir, obj_name, paste0(obj_name, '_', assay_name, '_genes.tsv')),
  sep='\t',
  col.names=TRUE,
  row.names=TRUE
)
```
